[DataPipe] Convert MapDataPipe.shuffle to IterDataPipe

ejguan opened the pull request:
Fixes: https://github.com/pytorch/data/issues/718

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #82975 [DataLoader] Add shuffling logic for MapDataPipe
* **#83202 [DataPipe] Convert MapDataPipe.shuffle to IterDataPipe**


This is an alternative PR against https://github.com/pytorch/pytorch/pull/82974

This PR would change the behavior for both types to the same behavior as `IterDataPipe.shuffle`
- Lazily generating seed per iteration
- Each iterators has a new seed
- Convert `MapDataPipe.shuffle` to an `IterDataPipe`

## BC-breaking Note:
This PR changes the return type of `MapDataPipe.shuffle` from a `MapDataPipe` to a `IterDataPipe`.

### 1. 12
Output as `MapDataPipe`
```
>>> from torch.utils.data import IterDataPipe, MapDataPipe
>>> from torch.utils.data.datapipes.map import SequenceWrapper
>>> dp = SequenceWrapper(list(range(10))).shuffle()
>>> isinstance(dp, MapDataPipe)
True
>>> isinstance(dp, IterDataPipe)
False
```

### This PR:
Output as `IterDataPipe`
```
>>> from torch.utils.data import IterDataPipe, MapDataPipe
>>> from torch.utils.data.datapipes.map import SequenceWrapper
>>> dp = SequenceWrapper(list(range(10))).shuffle()
>>> isinstance(dp, MapDataPipe)
False
>>> isinstance(dp, IterDataPipe)
True
```

@diff-train-skip-merge

ejguan committed:
[DataPipe] Convert MapDataPipe.shuffle to IterDataPipe

[ghstack-poisoned]

ejguan committed:
Update on "[DataPipe] Convert MapDataPipe.shuffle to IterDataPipe"




[ghstack-poisoned]

ejguan committed:
Update on "[DataPipe] Convert MapDataPipe.shuffle to IterDataPipe"




[ghstack-poisoned]

NivekT commented on torch/utils/data/datapipes/iter/combinatorics.py:
```diff
@@ -116,12 +116,15 @@ def set_shuffle(self, shuffle=True):
             for x in self.datapipe:
                 yield x
         else:
+            if self._seed is None:
+                self._seed = int(torch.empty((), dtype=torch.int64).random_().item())
```
1. If I understand this implementation correctly - even if the user doesn't set seed for every iteration, a new seed will still be created and used.
2. I actually think the behavior may not be different from the existing implementation where a new seed is created during `reset()`, which I believe is triggered lazily as well. Not 100% confident though.
3. This is in contrast with the other PR that keeps the seed fix unless the `set_seed` is called. Any reason why you changed your mind?

ejguan replied:
I am changing it because `reset` is not invoked during the state of `NotStarted`. (It was a bug and fixed by this PR)

> This is in contrast with the other PR that keeps the seed fix unless the set_seed is called. Any reason why you changed your mind?

The other PR keeps seed fix unless the set_seed is called because that's the only way to align the behavior between `MapDataPipe.shuffle` and `IterDataPipe.shuffle`, when `MapDataPipe.shuffle` is still a `MapDataPipe`. Let's talk about it in the meeting.

ejguan replied:
Having different seeds is preferred. But, the other PR is a compromisation for the sake of behavioral alignment.

NivekT commented on torch/utils/data/datapipes/map/combinatorics.py:
```diff
@@ -44,23 +46,58 @@ def __init__(self,
+
     def __iter__(self) -> Iterator[T_co]:
-        for i in self.indices:
-            yield self.datapipe[i]
+        if not self._enabled:
+            for idx in self.indices:
+                yield self.datapipe[idx]
+        else:
+            if self._seed is None:
+                self._seed = int(torch.empty((), dtype=torch.int64).random_().item())
+            self._rng.seed(self._seed)
+            self._seed = None
+            shuffled_indices = self._rng.sample(self.indices, len(self.indices))
+            for idx in shuffled_indices:
+                yield self.datapipe[idx]
```
It seems like this is conceptually applying `RandomSampler` to a `MapDataPipe` and thus making it a `IterDataPipe`. That is fine.

I do worry about losing random access to the samples. I suspect users might care about that? Is there anything that users can't do if we convert this to `IterDataPipe`?

If so, we probably want to have `Shuffler` and `RandomSamplerDataPipe` separate.

ejguan replied:
Let's consider a non-0-based index `MapDataPipe`. {"a": 1, "b": 2, "c": 3}
You want to access an element with index "a", do you want `MapDataPipe.shuffle` returns value 1 or any other value? Returning 1 makes shuffling non-functional; returning any other value makes users confused.
And, the order concept for `MapDataPipe` is tied to the order of indices. `__getitem__` function itself should not get involve any order.

Right now, I have two options in mind for `MapDataPipe.shuffle`:
- Like this PR, converting to an `IterDataPipe`
- Keep it as a `MapDataPipe`, but `__getitem__` gives non-shuffled value and `__iter__` gives results based on shuffled indices.

And, only with the above options, can I align the behavior of two types of `DataPipe` to always use different seeds per iteration, which is preferred IMHO.

NivekT replied:
It is not 100% clear to me what users expect when shuffle a custom non-0-index. Do they want the valued shuffled? Or do they just want the ordering returned from `__iter__` to change? I guess we can get rid of the ambiguity can strictly doing one over the other. My take is that users more likely want the ordering to change because the swapping the values among keys isn't that helpful?

> `__getitem__` gives non-shuffled value

This is likely confusing to users, and it is especially problematic if users chain another MapDataPipe operation (e.g. `.batch`) after `.shuffle()` because practically there will be no shuffling done. At that point, making it `IterDataPipe` is better because it should prevent unexpected behaviors.

This is all assuming that users do not want to swap the values among keys when they call `.shuffle()`.

VitalyFedyunin commented on torch/utils/data/datapipes/map/combinatorics.py:
```diff
@@ -44,23 +46,58 @@ def __init__(self,
+        return state
+
+    def __setstate__(self, state):
+        (
```
What will happen if you save and restore mid-way. I don't see any use of `_number_of_samples_yielded` or saving of `shuffled_indices`

ejguan replied:
Yeah. I should put `shuffled_indices` into the state.

VitalyFedyunin replied:
Please include test as well.

NivekT commented on torch/utils/data/datapipes/map/combinatorics.py:
```diff
@@ -1,17 +1,16 @@
 
-@functional_datapipe('shuffle')
-class ShufflerMapDataPipe(MapDataPipe[T_co]):
+class ShufflerIterDataPipe(IterDataPipe[T_co]):
```
```suggestion
# @functional_datapipe('shuffle')
class ShufflerIterDataPipe(IterDataPipe[T_co]):
```
Sorry for the annoying addition.

We need that extra commented out line in order to automatically generate the method interface for `MapDataPipe`.

Then we need to replace line 205 of `torch/utils/data/datapipes/gen_pyi.py` from:
`mapDP_method_to_special_output_type: Dict[str, str] = {}`
To:
`mapDP_method_to_special_output_type: Dict[str, str] = {"shuffle": "IterDataPipe"}`

ejguan replied:
Thanks for mentioning it! Will add it.

This might be another BE task for PYI gen. lol

NivekT replied:
Yea, the most ideal case is for `register_datapipe_as_function` to handle it.

At the same time, we generally don't plan to add more DataPipes (especially ones that require special treatments), so the cost/benefit tradeoff makes it very low priority.

NivekT reviewed (commented):
Small changes are needed for interface generation.

 Please update the MapDataPipe version's docstring so that users know it will be converted to `IterDataPipe` as well.

Question: Will the duplicate name `ShufflerIterDataPipe` cause any problem? I can't think of any since they are separated into different directories. I think it should be fine.

ejguan commented on torch/utils/data/datapipes/iter/combinatorics.py:
```diff
@@ -116,12 +116,15 @@ def set_shuffle(self, shuffle=True):
             for x in self.datapipe:
                 yield x
         else:
+            if self._seed is None:
+                self._seed = int(torch.empty((), dtype=torch.int64).random_().item())
```
@NivekT There is a separate problem for `reset`.

As I said about `reset`, it is not invoked when the datapipe state is `NotStarted`. So, I have to add the manual reset logic here to generate `seed`. But, it prevents this DataPipe  not resetting after deserialization unless I add an extra flag to indicate if the datapipe is restored or not (this is contrary to our idea to use `reset` by default).

ejguan replied:
I we do so, then I will remove the manual reset logic here.

NivekT replied:
I think the extra reset may be somewhat unnecessary for most DataPipes in the beginning. However, I don't think it should be noticeable and should not break anything. I just tried looking through if it will causes any edge cases for cases with ChildDataPipes but I can't think of any. You can make that change and confirm.

ejguan committed:
Update on "[DataPipe] Convert MapDataPipe.shuffle to IterDataPipe"

Fixes: https://github.com/pytorch/data/issues/718



This is an alternative PR against https://github.com/pytorch/pytorch/pull/82974

This PR would change the behavior for both types to the same behavior as `IterDataPipe.shuffle`
- Lazily generating seed per iteration
- Each iterators has a new seed
However, this requires `MapDataPipe.shuffle` is converted as a `IterDataPipe`. I prefer this option simply because I find the concept of shuffling `MapDataPipe` would be shuffling indices and returning the data in the order of shuffled indices.

[ghstack-poisoned]

ejguan committed:
Update on "[DataPipe] Convert MapDataPipe.shuffle to IterDataPipe"

Fixes: https://github.com/pytorch/data/issues/718



This is an alternative PR against https://github.com/pytorch/pytorch/pull/82974

This PR would change the behavior for both types to the same behavior as `IterDataPipe.shuffle`
- Lazily generating seed per iteration
- Each iterators has a new seed
However, this requires `MapDataPipe.shuffle` is converted as a `IterDataPipe`. I prefer this option simply because I find the concept of shuffling `MapDataPipe` would be shuffling indices and returning the data in the order of shuffled indices.

[ghstack-poisoned]

ejguan committed:
Update on "[DataPipe] Convert MapDataPipe.shuffle to IterDataPipe"

Fixes: https://github.com/pytorch/data/issues/718



This is an alternative PR against https://github.com/pytorch/pytorch/pull/82974

This PR would change the behavior for both types to the same behavior as `IterDataPipe.shuffle`
- Lazily generating seed per iteration
- Each iterators has a new seed
However, this requires `MapDataPipe.shuffle` is converted as a `IterDataPipe`. I prefer this option simply because I find the concept of shuffling `MapDataPipe` would be shuffling indices and returning the data in the order of shuffled indices.

[ghstack-poisoned]

ejguan committed:
Update on "[DataPipe] Convert MapDataPipe.shuffle to IterDataPipe"

Fixes: https://github.com/pytorch/data/issues/718



This is an alternative PR against https://github.com/pytorch/pytorch/pull/82974

This PR would change the behavior for both types to the same behavior as `IterDataPipe.shuffle`
- Lazily generating seed per iteration
- Each iterators has a new seed
- Convert `MapDataPipe.shuffle` to an `IterDataPipe`

[ghstack-poisoned]

NivekT commented on test/test_datapipe.py:
```diff
@@ -1813,6 +1830,16 @@ def test_shuffler_mapdatapipe(self):
+            exp.append(next(it))
+        shuffler_dp_copy = pickle.loads(pickle.dumps(shuffler_dp))
+        exp.extend(list(it))
+        self.assertEqual(exp, list(shuffler_dp_copy))
```
Similar to the last PR, this test needs an update as we expect fewer elements from the copy.

NivekT reviewed (commented):
Overall LGTM! We have the questions as the last one; will approve as soon as we finalize the details on the last one and modify this one the similarly.

ejguan committed:
Update on "[DataPipe] Convert MapDataPipe.shuffle to IterDataPipe"

Fixes: https://github.com/pytorch/data/issues/718



This is an alternative PR against https://github.com/pytorch/pytorch/pull/82974

This PR would change the behavior for both types to the same behavior as `IterDataPipe.shuffle`
- Lazily generating seed per iteration
- Each iterators has a new seed
- Convert `MapDataPipe.shuffle` to an `IterDataPipe`

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/gen_pyi.py:
```diff
@@ -202,7 +202,7 @@ def get_method_definitions(file_path: Union[str, List[str]],
 mapDP_files_to_exclude: Set[str] = {"__init__.py", "utils.py"}
 mapDP_deprecated_files: Set[str] = set()
-mapDP_method_to_special_output_type: Dict[str, str] = {}
+mapDP_method_to_special_output_type: Dict[str, str] = {"shuffle": "IterDataPipe"}
```
Updated to make sure it works as expected

ejguan committed:
Update on "[DataPipe] Convert MapDataPipe.shuffle to IterDataPipe"

Fixes: https://github.com/pytorch/data/issues/718



This is an alternative PR against https://github.com/pytorch/pytorch/pull/82974

This PR would change the behavior for both types to the same behavior as `IterDataPipe.shuffle`
- Lazily generating seed per iteration
- Each iterators has a new seed
- Convert `MapDataPipe.shuffle` to an `IterDataPipe`

[ghstack-poisoned]

NivekT reviewed (approved):
The implementation LGTM! Just need to update the docstring. Some ideas:
1. Mention that `IterDataPipe` will now be returned (and possibly why and use converter to get `MapDataPipe` back again) - you can decide what to mention
2. In the example, call `list(shuffle_dp)` once more to show that the result will be different in the second call
3. We can potentially mention the `set_seed` API but if we expect users to rely on `DataLoader` then maybe not

ejguan committed:
Update on "[DataPipe] Convert MapDataPipe.shuffle to IterDataPipe"

Fixes: https://github.com/pytorch/data/issues/718



This is an alternative PR against https://github.com/pytorch/pytorch/pull/82974

This PR would change the behavior for both types to the same behavior as `IterDataPipe.shuffle`
- Lazily generating seed per iteration
- Each iterators has a new seed
- Convert `MapDataPipe.shuffle` to an `IterDataPipe`

## BC-breaking Note:
This PR changes the return type of `MapDataPipe.shuffle` from a `MapDataPipe` to a `IterDataPipe`.

### 1. 12
Output as `MapDataPipe`
```
>>> from torch.utils.data import IterDataPipe, MapDataPipe
>>> from torch.utils.data.datapipes.map import SequenceWrapper
>>> dp = SequenceWrapper(list(range(10))).shuffle()
>>> isinstance(dp, MapDataPipe)
True
>>> isinstance(dp, IterDataPipe)
False
```

### This PR:
Output as `IterDataPipe`
```
>>> from torch.utils.data import IterDataPipe, MapDataPipe
>>> from torch.utils.data.datapipes.map import SequenceWrapper
>>> dp = SequenceWrapper(list(range(10))).shuffle()
>>> isinstance(dp, MapDataPipe)
False
>>> isinstance(dp, IterDataPipe)
True
```

[ghstack-poisoned]

ejguan commented:
@pytorchbot merge -g

weiwangmeta commented:
@pytorchbot merge -l
